tensorflow/compiler/xla/service/slice_delaying.cc

```diff
@@ -0,0 +1,295 @@
+        if (slice_j->opcode() != HloOpcode::kSlice ||
+            slice_j->operand(0) != operands[j] ||
+            slice_0->slice_starts() != slice_j->slice_starts() ||
+            slice_0->slice_limits() != slice_j->slice_limits() ||
+            slice_0->slice_strides() != slice_j->slice_strides()) {
+          isValidUser = false;
+          break;
+        }
+      }
+      if (!isValidUser) {
+        continue;
+      }
+
+      // found the user
+      users.push_back(user);
+      visited_.insert(user);
+    }  // end for loop slice_0->users
+  }  // end for loop operand0's slice users
+
+  // calculate the cost. If the no user found or only few small users, skip this
+  // instruction.
+  if (users.empty()) {
+    return tensorflow::errors::FailedPrecondition(
+        "No found valid users");
+  } else if (!ShouldReplace(operands, users)) {
+    return tensorflow::errors::FailedPrecondition(
+        "No Enough elements slice");
+  } else {
+    return users;
+  }
+}
+
+void SliceDelayer::GenerateNewOp(const std::vector<HloInstruction*>& operands,
+    const std::vector<HloInstruction*>& users) {
+  // generate new ops
+  const Shape shape = operands[0]->shape();
+  HloComputation* computation = users[0]->parent();
+  auto new_op = computation->AddInstruction(
+      users[0]->CloneWithNewOperands(shape, operands));
+  VLOG(10) << "Add NewOp: " << new_op->ToString();
+
+  // replace the old user and its slice operands with new operation and its
+  // slice user.
+  for (HloInstruction* user : users) {
+    const HloInstruction* slice = user->operand(0);
+    auto new_user = computation->AddInstruction(
+        slice->CloneWithNewOperands(user->shape(), {new_op}));
+    VLOG(10) << "Add NewSlice: " << new_user->ToString()
+            << "\nReplace: " << user->ToString();
+    user->ReplaceAllUsesWith(new_user);
+
+    for (HloInstruction* slice_operand : user->operands()) {
+      slices_.insert(slice_operand);
+    }
+    removed_.insert(user);
+  }  // end for users
+}
+
+StatusOr<bool> SliceDelaying::Run(HloModule* module) {
+  VLOG(3) << "Run Pass: " << name();
+  VLOG(10) << "before: " << name() << "\n" << module->ToString();
+  SliceDelayer slice_delayer;
+  bool changed = false;
+
+  for (HloComputation* computation : module->computations()) {
+    for (HloInstruction* instruction :
+        computation->MakeInstructionPostOrder()) {
+      // Skip the visited instruction tries merge elementwise instruction with
+      // its peers.
+      if (!slice_delayer.IsVisited(instruction) &&
+          instruction->IsElementwise() && instruction->operand_count() != 0) {
+        // TODO(xinan): more other instructions
+        VLOG(10) << "Merge inst: " << instruction->ToString();
+        changed |= slice_delayer.MergeWithPeers(instruction).ok();
+      }
+    }  // end for instructions in computation
+  }  // end for computations in module
+
+  // Clears dead nodes
+  slice_delayer.EliminateDeadInstructions();
```

Comment:
Yes, this function is similar as DCE pass. It eliminates dead code which generates in this pass. 
Do you think that it is better to leave the dead code to another DCE pass rather than to handle it in the pass?

Reply:
> Do you think that it is better to leave the dead code to another DCE pass rather than to handle it in the pass?

Yes.  This is how most other passes work in XLA.  If you look at e.g. nvptx_compiler, this pass would be part of the "simplification pipeline", which includes DCE and runs to a fixed point, i.e. until no changes are made.  So there would be a loop in the compiler that runs both this pass and DCE.